dm/syncer/error.go

```diff
@@ -61,6 +65,139 @@ func isDropColumnWithIndexError(err error) bool {
 			strings.Contains(mysqlErr.Message, "with tidb_enable_change_multi_schema is disable"))
 }
 
+// getDDLStatusFromTiDB retrieves the synchronizing status of DDL from TiDB
+// hence here db should be TiDB database.
+func getDDLStatusFromTiDB(ctx context.Context, db *sql.DB, ddl string, createTime int64) (string, error) {
+	rowNum := 10
+	count := 0
+	rowOffset := 0
+
+	showJobsLimit := fmt.Sprintf("ADMIN SHOW DDL JOB QUERIES LIMIT 10 OFFSET %d", rowOffset)
+	rowsLimit, err := db.QueryContext(ctx, showJobsLimit)
+	if err != nil {
+		rowsLimit.Close()
+		return "", err
+	}
+
+	var columnsLimit []string
+	columnsLimit, err = rowsLimit.Columns()
+	if err != nil {
+		rowsLimit.Close()
+		return "", err
+	}
+
+	var resultsLimit [][]string
+	resultsLimit, err = export.GetSpecifiedColumnValuesAndClose(rowsLimit, columnsLimit...)
+	if err != nil {
+		rowsLimit.Close()
+		return "", err
+	}
+
+	for {
+		// every attempt try 10 history jobs
+		showJobs := fmt.Sprintf("ADMIN SHOW DDL JOBS %d", rowNum)
+		rows, err := db.QueryContext(ctx, showJobs)
+		if err != nil {
+			rows.Close()
+			return "", err
+		}
+
+		var columns []string
+		columns, err = rows.Columns()
+		if err != nil {
+			rows.Close()
+			return "", err
+		}
+
+		var results [][]string
+		results, err = export.GetSpecifiedColumnValuesAndClose(rows, columns...)
+		if err != nil {
+			rows.Close()
+			return "", err
+		}
+
+		for i := rowNum - 10; i < rowNum; i++ {
+			ddlCreateTimeStr := results[i][8]
+			var ddlCreateTimeParse time.Time
+			ddlCreateTimeParse, err = time.Parse("2006-01-02 15:04:05", ddlCreateTimeStr)
+			if err != nil {
+				return "", err
+			}
+			ddlCreateTime := ddlCreateTimeParse.Unix()
+
+			// ddlCreateTime and createTime are both based on timezone of downstream
+			if ddlCreateTime >= createTime {
+				var jobID int
+				jobID, err = strconv.Atoi(results[i][0])
+				if err != nil {
+					return "", err
+				}
+
+				offset := rowNum + count - 10
+				for {
+					flag := false
+					currentOffset := offset
+					for {
+						if currentOffset == len(resultsLimit)-1 {
+							// expand resultsLimit for deeper search
+							rowOffset += 10
+							showJobsLimitNext := fmt.Sprintf("ADMIN SHOW DDL JOB QUERIES LIMIT 10 OFFSET %d", rowOffset)
+							var rowsLimitNext *sql.Rows
+							rowsLimitNext, err = db.QueryContext(ctx, showJobsLimitNext)
+							if err != nil {
+								rowsLimitNext.Close()
+								return "", err
+							}
+
+							var columnsLimitNext []string
+							columnsLimitNext, err = rowsLimitNext.Columns()
+							if err != nil {
+								rowsLimitNext.Close()
+								return "", err
+							}
+
+							var resultsLimitNext [][]string
+							resultsLimitNext, err = export.GetSpecifiedColumnValuesAndClose(rowsLimitNext, columnsLimitNext...)
+							if err != nil {
+								rowsLimitNext.Close()
+								return "", err
+							}
+
+							for k := 0; k < 10; k++ {
+								resultsLimit = append(resultsLimit, resultsLimitNext[k])
```

Comment:
resultsLimit = append(resultsLimit, resultsLimitNext...)

And downstream may get another DDL between two queries, so we should not directly concatenate them